File: 125-valid-palindrome/valid-palindrome.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        int n = s.size();
        int left = 0;
        int right = n-1;

        while(left < right) {
            while(left < right && !isalnum(s[left])) left++;
            while(right > left && !isalnum(s[right])) right--;
            if(left < right) {
                if(tolower(s[left]) != tolower(s[right])) return false;
            }
            left++;
            right--;
        }

        return true;
    }
};

File: 125-valid-palindrome/valid-palindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        StringBuilder sb = new StringBuilder();
        for(char ch: s.toCharArray()) {
            if(Character.isLetterOrDigit(ch)) {
                sb.append(Character.toLowerCase(ch));
            }
        }

        return checkPalindrome(sb);
    }

    private boolean checkPalindrome(StringBuilder s) {
        int n = s.length();
        for(int i=0; i < n/2; i++) {
            if(s.charAt(i) != s.charAt(n-i-1)) return false;
        }

        return true;
    }
}